### PR TITLE
Use credentials for update

### PR DESCRIPTION
**Pull request by gabyx**

### CODE REVIEW

**Comment by gabyx** [comment]
@rycus86 : There is only one test failing still: `test-rules` where do I find the entry to this test? I could not find it in the travis on my branch?

**Comment by rycus86** [comment]
It's in https://github.com/rycus86/githooks/blob/master/tests/test-rules.sh which runs most of the pre-commit hooks that should have run before CI already.
And the logs for it are here for your branch: https://travis-ci.org/rycus86/githooks/jobs/555685071
Looks like `shfmt` formatting issues only if I'm reading it right (on a mobile atm).

**Comment by gabyx** [comment]
Finally. 😓

**Comment by gabyx** [comment]
@rycus86 : Would be cool if you could have a look at the changes, its no so much. Should we try to increase the Coverage with some more Tests?

**Comment by rycus86** [comment]
> @rycus86 : Would be cool if you could have a look at the changes, its no so much. Should we try to increase the Coverage with some more Tests?

OK, let me go through it quickly. I'm not too worried about tests for now, we can sort them out later.

File: base-template.sh [hunk 1]
```diff
@@ -545,7 +611,7 @@ fetch_latest_update_script() {
 #   None
 #####################################################
 read_updated_version_number() {
-    LATEST_VERSION=$(echo "$INSTALL_SCRIPT" | grep "^# Version: .*" | head -1 | sed "s/^# Version: //")
+    LATEST_VERSION=$(grep "^# Version: .*" <"$INSTALL_SCRIPT" | head -1 | sed "s/^# Version: //")
```
[the remark this answers is not in the document]
Comment (rycus86) [reply base-template.sh]: Also `sed -E` please

File: cli.sh [hunk 2]
```diff
@@ -34,6 +34,7 @@ Available commands:
     readme      Manages the Githooks README in the current repository
     ignore      Manages Githooks ignore files in the current repository
     config      Manages various Githooks configuration
+    apps        Manages script folders for apps
```
Comment (rycus86) [inline cli.sh]: `tools` instead of `apps`?

File: cli.sh [hunk 3]
```diff
@@ -51,6 +52,56 @@ print_help_header() {
...
+#####################################################
+# Set the current install dir.
+# This script location determines it.
+#
+# Sets the ${INSTALL_DIR} variable.
+#
+# Returns:
+#   0 if suncessful, 1 if not
+#####################################################
+set_install_dir() {
```
Comment (rycus86) [inline cli.sh]: Let's bin `set_install_dir`, `make_abs_path` and `is_abs_path` for now, and just go with `~/.githooks/tools`

File: cli.sh [hunk 4]
```diff
@@ -1270,7 +1323,7 @@ git hooks update [enable|disable]
 
     if [ "$1" != "force" ]; then
         if ! is_update_available; then
-            echo "  Githooks is already on the latest version"
+            echo "  Current Githooks version: $CURRENT_VERSION is newer than the remote version: $LATEST_VERSION"
```
Comment (rycus86) [inline cli.sh]: Can we revert this change? I don't think the extra information is useful for most people.

File: cli.sh [hunk 5]
```diff
@@ -1297,30 +1351,75 @@ record_update_time() {
...
 #
 # Returns:
-#   1 if failed the load the script, 0 otherwise
+#   0 and "$INSTALL_DIR/apps/$1/run.sh"
+#   1 and "" otherwise
 #####################################################
-fetch_latest_install_script() {
-    DOWNLOAD_URL="https://raw.githubusercontent.com/rycus86/githooks/master/install.sh"
+get_app_run_script() {
+    if [ -d "$INSTALL_DIR" ] && [ -f "$INSTALL_DIR/apps/$1/run.sh" ]; then
```
Comment (rycus86) [inline cli.sh]: Same comments as in the other file.

File: cli.sh [hunk 6]
```diff
@@ -1468,27 +1567,14 @@ git hooks readme [add|update]
...
-
-    else
-        echo "! Failed to fetch the latest README - needs either curl or wget"
-        return 1
-    fi
-
-    # shellcheck disable=SC2181
-    if [ $? -ne 0 ]; then
+    README_FILE="$(mktemp)"
+    if ! download_file "README.md" "$README_FILE"; then
```
Comment (rycus86) [inline cli.sh]: This file here needs to refer to `.githooks/README.md` not the main `README.md`

File: cli.sh [hunk 7]
```diff
@@ -1897,6 +1983,126 @@ config_update_last_run() {
     fi
 }
 
+#####################################################
+# Manages the app script folders.
+#
+# Returns:
+#   1 on failure, 0 otherwise
+#####################################################
+manage_apps() {
```
Comment (rycus86) [inline cli.sh]: `apps` to `tools` in this function

File: cli.sh [hunk 8]
```diff
@@ -1897,6 +1983,126 @@ config_update_last_run() {
...
+    if [ "$1" = "help" ]; then
+        print_help_header
+        echo "
+git hooks apps install download
```
Comment (rycus86) [inline cli.sh]: I'd change this to `git hooks tools register <type> <script>` where `<type>` would be `download` or similar here, and you could point to any file with `<script>` that gets copied into the global folder we want it in.

File: tests/exec-steps.sh [hunk 9]
```diff
@@ -39,7 +39,9 @@ for STEP in /var/lib/tests/step-*.sh; do
     elif [ $TEST_RESULT -ne 0 ]; then
         FAILURE=$(echo "$TEST_OUTPUT" | tail -1)
         echo "! $STEP has failed with code $TEST_RESULT ($FAILURE), output:"
+        echo "--------"
```
Comment (rycus86) [inline tests/exec-steps.sh]: Are these necessary? Can you please revert?

File: tests/step-029.sh [hunk 10]
```diff
@@ -2,7 +2,7 @@
 # Test:
 #   Direct template execution: execute auto-update
 
-if ! curl --version && ! wget --version; then
+if ! curl --version >/dev/null 2>&1 && ! wget --version >/dev/null 2>&1; then
```
Comment (rycus86) [inline tests/step-029.sh]: Not sure why this needs to change? I think it's good to have their version printed when available, so it's easier to see when it goes wrong. If the test is successful, their output will be hidden anyway.

File: tests/step-032.sh [hunk 11]
```diff
@@ -2,7 +2,7 @@
 # Test:
 #   Direct template execution: choose to ignore the update
 
-if ! curl --version && ! wget --version; then
+if ! curl --version >/dev/null 2>&1 && ! wget --version >/dev/null 2>&1; then
```
Comment (rycus86) [inline tests/step-032.sh]: Same as above, please revert this.

File: tests/step-042.sh [hunk 12]
```diff
@@ -2,7 +2,7 @@
 # Test:
 #   Run a single-repo install and try the auto-update
 
-if ! curl --version && ! wget --version; then
+if ! curl --version >/dev/null 2>&1 && ! wget --version >/dev/null 2>&1; then
```
Comment (rycus86) [inline tests/step-042.sh]: Same as above, please revert this.

File: tests/step-062.sh [hunk 13]
```diff
@@ -2,7 +2,7 @@
 # Test:
 #   Cli tool: run forced update
 
-if ! curl --version && ! wget --version; then
+if ! curl --version >/dev/null 2>&1 && ! wget --version >/dev/null 2>&1; then
```
Comment (rycus86) [inline tests/step-062.sh]: Same as above, please revert this.

File: tests/step-063.sh [hunk 14]
```diff
@@ -2,7 +2,7 @@
 # Test:
 #   Cli tool: run an update
 
-if ! curl --version && ! wget --version; then
+if ! curl --version >/dev/null 2>&1 && ! wget --version >/dev/null 2>&1; then
```
Comment (rycus86) [inline tests/step-063.sh]: Same as above, please revert this.

File: tests/step-076.sh [hunk 15]
```diff
@@ -2,7 +2,7 @@
 # Test:
 #   Direct template execution: choose to ignore the update (non-single)
 
-if ! curl --version && ! wget --version; then
+if ! curl --version >/dev/null 2>&1 && ! wget --version >/dev/null 2>&1; then
```
Comment (rycus86) [inline tests/step-076.sh]: Same as above, please revert this.

File: tests/step-077.sh [hunk 16]
```diff
@@ -2,7 +2,7 @@
 # Test:
 #   Direct template execution: choose to ignore the update (single)
 
-if ! curl --version && ! wget --version; then
+if ! curl --version >/dev/null 2>&1 && ! wget --version >/dev/null 2>&1; then
```
Comment (rycus86) [inline tests/step-077.sh]: Same as above, please revert this.

File: tests/step-080.sh [hunk 17]
```diff
@@ -2,7 +2,7 @@
 # Test:
 #   Cli tool: add/update README
 
-if ! curl --version && ! wget --version; then
+if ! curl --version >/dev/null 2>&1 && ! wget --version >/dev/null 2>&1; then
```
Comment (rycus86) [inline tests/step-080.sh]: Same as above, please revert this.

File: tests/step-094.sh [hunk 18]
```diff
@@ -2,7 +2,7 @@
 # Test:
 #   Cli tool: run an installation
 
-if ! curl --version && ! wget --version; then
+if ! curl --version >/dev/null 2>&1 && ! wget --version >/dev/null 2>&1; then
```
Comment (rycus86) [inline tests/step-094.sh]: Same as above, please revert this.

File: tests/step-100.sh [hunk 19]
```diff
@@ -23,7 +23,7 @@ if grep -r 'github.com/rycus86/githooks' ~/test100/; then
     exit 1
 fi
 
-# run the install, and let it intall into existing repos
+# run the install, and let it install into existing repos
```
Comment (rycus86) [inline tests/step-100.sh]: Thanks for this!

File: tests/test-rules.sh [hunk 20]
```diff
@@ -20,7 +20,6 @@ run_pre_commit_test() {
     fi
 }
 
-run_pre_commit_test no-single-quote
```
Comment (rycus86) [inline tests/test-rules.sh]: Put this one back please.

**Comment by gabyx** [comment]
Thx for the comments:
@rycus86 : are you really keen on changing back to single quotes only  ' '  ( I can do this), do you also want the HEREDOC reverted? I find it so much more flexibel and also its much more safe to use backticks in shell scripts and only if needed refere to double quotes which are always expanding, which is not always what we want, especially for regexes etc... Just give me a yes or no, than I revert it ;)

**Comment by rycus86** [comment]
If you can, then yes please. I think they make it harder to make mistakes because you need to think harder if it's going to do what you want or not. :)

Thanks again and sorry for being a bit slow with this, I think we're in very different timezones that doesn't help too much.

File: tests/step-029.sh (hunk 10 above)
Comment (gabyx) [reply tests/step-029.sh]: True!

File: tests/exec-steps.sh (hunk 9 above)
Comment (rycus86) [reply tests/exec-steps.sh]: Thanks :)

File: cli.sh (hunk 3 above)
Comment (rycus86) [reply cli.sh]: SLOA?

File: cli.sh (hunk 4 above)
Comment (rycus86) [reply cli.sh]: Yeah, I think it's useful when working on update related changes, but not necessarily that useful otherwise.

File: cli.sh (hunk 8 above)
Comment (rycus86) [reply cli.sh]: Yeah, I suppose that makes sense. But is this something you actually need at the moment? I'd start with a simple single file for now, and add extra bits only if needed. Codewise it's not much of a difference, but too many options can get confusing to people.

File: cli.sh [hunk 21]
```diff
@@ -1297,30 +1328,93 @@ record_update_time() {
...
+#   script into a variable.
+#
+# Sets the ${INSTALL_SCRIPT} variable
+#
+# Returns:
+#   1 if failed the load the script, 0 otherwise
+#####################################################
+fetch_latest_install_script() {
+    INSTALL_SCRIPT="$(mktemp)"
+    # shellcheck disable=SC2181
```
[the remark this answers is not in the document]
Comment (gabyx) [reply cli.sh]: Thx, refactoring error.

**Comment by gabyx** [comment]
Now it should be good to go.

**Comment by rycus86** [comment]
Thanks a lot for all the changes @gabyx !
It looks good to me, I think it's ready to merge. :tada: 

Can I ask you to please add `(experimental)` to the new headers in the readme files for now, then I'll try to add some tests to make sure each system we test on is happy with it, then I'll drop the experimental label.

Also, once you've done that, please squash your commits into one and force push, then I'll merge it in.
Thanks!

**Comment by gabyx** [comment]
Done. Lets see if the tests work.

**Comment by rycus86** [comment]
Thanks a lot again @gabyx !
